# Transcriptomic integration of skin cancer datasets
This repository contains the R scripts used for the integration of multiple microarray and RNA-seq datasets coming from NCBI GEO and ArrayExpress web repositories, helping to determine those multiclass *Differentially Expressed Genes (DEGs)* related to multiple skin cancer states. The different R scripts should be sequentially run step-by-step and in the next order:

# 1) RawDataAcquisition_and_Preprocessing.R 
This script starts saving the current working directory. The user should indicate what working directory will use to build the folder system during this analysis. Next, different R libraries are needed after using **biocLite.R** script:
```
source("http://bioconductor.org/biocLite.R")
```
Then, the microarray analysis starts by acquiring (using ```GEOquery```) and preprocessing different GEO series. For each series, different quality analysis iterations can be needed in order to eventually discard detected outliers (those samples presenting low quality for any assessed quality test included in ```arrayQualityMetrics```). Once this is done, RMA normalization has to be applied for those unpreprocessed series previously (using ```affy```,```oligo``` and ```lumi```). Additionally, sample labels of those selected samples have to be stored as well. The application of a gene common annotation (using ```Ã nnotate```) is recommended at this point and required for our analysis, thinking in the further dataset integration. Each series has to be mapped from manufacturer chip identifiers (such as ```hgu133a.db```,```hgu133plus2.db```, etc.) to standardized symbols (such as HUGO, EntrezID, etc.). With respect to RNA-seq series, this script performs from counts previously obtained to equivalent gene expression values. Gene annotation is obtained by retrieving from ```BiomaRt``` databases. GC content bias is also corrected by means of ```cqn```, computing the normalized and equivalent gene expression values.

# 2) GeneExpressionIntegration.R
This script continues to analyse those datasets previously acquired and preprocessed in the above point (if the user has available samples quantified in gene expression values, it is possible to continue from here as well). Firstly, it is necessary to summarize all those genes transcribing the same identifier (commonly, you can apply summarization by means or median). Next, logarithmic transformation has to be applied for those series whose depth is not log2 scaled. After this, the scale depth of each manufacturer chip must be inspected, checking representation similarities for different series using the same chip identifier. The user can choice a specific bit depth to be established (for example, we have selected 16-bit depth to homogeneize all samples, partly due to most series were already presenting that depth). In order to work with exclusively real information, only complete cases were selected for the integration: those genes appearing in all selected series. After considering that, batch effect correction was performed by means of **ComBat** method (included in ```sva```). Finally, in order to avoid further erroneus diagnoses applying machine learning classifiers, an inter-array normalization is considered by means of ```normalizeBetweenArrays```. The integrated dataset is ready at this point.

# 3) OVOMulticlassDEGsSelection.R
This script allows to determine multiclass DEGs related to the different pathological states considered. In order to strengthen the process of identification and selection of potential reliable biomarkers, a robust cross-validation procedure was designed by separating in three datasets: training, validation and testing. Specifically, up to M = 10 experiments (different partitions of the integrated dataset) were considered where the feature selection process was additionally embedded and performed. Furthermore, similar representativeness of each pathological state was ensured within each partition in order to guarantee real possibilities to be well-diagnosed for those emerged multiclass DEGs. The consideration of mainly two parameters was particularly considered as relevant in order to determine those DEGs: log2 fold change (LFC) and maximum number of DEGs to distinguish among two specific pathological states (NMAX). The script is automated to obtain different DEGs set depending on established thresholds for both parameters: from LFC = 1 to 5 increasing 0.5 at each step, and from NMAX = 1 to 10 increasing 1 at each step. These values can be modified and assessed at the user's choice. Next, in order to remove eventual spurious DEGs (only appearing in specific experiments), all rankings belonging to the same parameter combination are matched and common genes are identified. At this point, these common genes are selected for each previous partition and reassessed. Additionally, the informative capability of those potential and identified multiclass DEGs can be assessed by applying feature selection algorithms such as *minimum Redundancy Maximum Relevance (mRMR)*, as here it is applied.

After reproducing all these scripts, the user could apply different Machine Learning techniques (such as *Support Vector Machines (SVMs), k-Nearest Neighbors (kNNs) or Naive Bayes (NB)* in order to assess the power of the multiclass DEGs for an intelligent diagnosis. To this end, we encourage to use our previously published **KnowSeq** R/Bioc package [1], thus additionally allowing to assess multiple functionalities such as *Gene Set Enrichment Analysis (GSEA)*.

**References**

[1] D. Castillo et al., "From RAW to Knowledge: The KnowSeq user guide," 2019.
